pymc3/distributions/multivariate.py

```diff
@@ -20,63 +20,79 @@
            'LKJCorr', 'LKJCholeskyCov', 'MatrixNormal']
 
 
-class _QuadFormBase(Continuous):
-    def __init__(self, mu=None, cov=None, chol=None, tau=None, lower=True,
-                 *args, **kwargs):
-        super(_QuadFormBase, self).__init__(*args, **kwargs)
-        if len(self.shape) > 2:
-            raise ValueError("Only 1 or 2 dimensions are allowed.")
+class _CovSet():
+    R"""
+    Convenience class to set Covariance, Inverse Covariance and Cholesky
+    descomposition of Covariance matrices.
+    """
+    def __initCov__(self, cov=None, tau=None, chol=None, lower=True):
+        if all([val is None for val in [cov, tau, chol]]):
+            raise ValueError('One of cov, tau or chol arguments must be provided.')
 
-        if chol is not None and not lower:
-            chol = chol.T
         if len([i for i in [tau, cov, chol] if i is not None]) != 1:
             raise ValueError('Incompatible parameterization. '
                              'Specify exactly one of tau, cov, '
                              'or chol.')
-        self.mu = mu = tt.as_tensor_variable(mu)
-        self.solve_lower = tt.slinalg.Solve(A_structure="lower_triangular")
-        # Step methods and advi do not catch LinAlgErrors at the
-        # moment. We work around that by using a cholesky op
-        # that returns a nan as first entry instead of raising
-        # an error.
-        cholesky = Cholesky(nofail=True, lower=True)
+
+        if chol is not None and not lower:
+            chol = chol.T
+
+        self.cov = self.tau = self.chol_cov = None
 
         if cov is not None:
-            self.k = cov.shape[0]
             self._cov_type = 'cov'
-            cov = tt.as_tensor_variable(cov)
-            if cov.ndim != 2:
-                raise ValueError('cov must be two dimensional.')
-            self.chol_cov = cholesky(cov)
-            self.cov = cov
-            self._n = self.cov.shape[-1]
-        elif tau is not None:
-            self.k = tau.shape[0]
-            self._cov_type = 'tau'
-            tau = tt.as_tensor_variable(tau)
-            if tau.ndim != 2:
-                raise ValueError('tau must be two dimensional.')
-            self.chol_tau = cholesky(tau)
-            self.tau = tau
-            self._n = self.tau.shape[-1]
-        else:
-            self.k = chol.shape[0]
+            self.cov = tt.as_tensor_variable(cov)
+        elif chol is not None:
             self._cov_type = 'chol'
-            if chol.ndim != 2:
-                raise ValueError('chol must be two dimensional.')
-            self.chol_cov = tt.as_tensor_variable(chol)
-            self._n = self.chol_cov.shape[-1]
+            self.chol = tt.as_tensor_variable(chol)
+            if self.chol.ndim != 2:
+                raise ValueError('`chol` must be two dimensional.')
+        else:
+            self._cov_type = 'tau'
+            self.tau = tt.as_tensor_variable(tau)
+
+class _QuadFormBase(Continuous, _CovSet):
```

Comment:
Even if `_CovSet` is an object, it should be an attribute of `_QuadFormBase`, not a superclass.